src/h5web/vis-packs/core/matrix/utils.ts

```diff
@@ -0,0 +1,23 @@
+import { hasComplexType, hasNumericType } from '../../../guards';
+import { renderComplex } from '../../../metadata-viewer/utils';
+import type {
+  ArrayShape,
+  Dataset,
+  H5WebComplex,
+} from '../../../providers/models';
+import type { PrintableType, ValueFormatter } from '../models';
+import { formatNumber } from '../utils';
+
+export function getFormatter(
+  dataset: Dataset<ArrayShape, PrintableType>
+): ValueFormatter<PrintableType> {
+  if (hasComplexType(dataset)) {
+    return (val) => renderComplex(val as H5WebComplex, '.2e');
```

Comment:
We don't bother asserting the array of values via the dtype any more (i.e. `isComplexArray(type: DType, arr: NdArray<unknown>): arr is NdArray<H5WebComplex>`) -- we just cast the value in each formatter to the type we know it has.

It's not perfect, but it's much simpler and much more understandable.

Reply:
> It's not perfect, but it's much simpler and much more understandable.

True, a well-struck blow :+1: